File: patterns/06_dependency_injection/java/Officium.java
// ============================================================
//  DEPENDENCY INJECTION — The Praetorian Quartermaster
//  Run: java Officium.java
// ============================================================
import java.util.*;

public class Officium {

    interface IMessenger { void send(String to, String msg); }
    interface ILogger    { void log(String event); }

    record ScrollMessenger() implements IMessenger {
        public void send(String to, String msg) {
            System.out.printf("  📜 Scroll to %s: '%s'%n", to, msg);
        }
    }
    static class MockMessenger implements IMessenger {
        final List<String[]> sent = new ArrayList<>();
        public void send(String to, String msg) {
            sent.add(new String[]{to, msg});
            System.out.printf("  🧪 MOCK: to=%s msg='%s'%n", to, msg);
        }
    }
    record ConsoleLogger() implements ILogger {
        public void log(String event) { System.out.println("  [LOG] " + event); }
    }

    static class MilitaryCampaign {
        private final IMessenger messenger;
        private final ILogger    logger;

        // Constructor injection
        MilitaryCampaign(IMessenger messenger, ILogger logger) {
            this.messenger = messenger;
            this.logger    = logger;
        }

        void launchAttack(String target) {
            logger.log("Campaign launched against " + target);
            messenger.send("Praetorian Guard", "ADVANCE on " + target + "!");
            messenger.send("Supply Corps",     "Provisions to " + target);
            logger.log("Orders dispatched");
        }
    }

    public static void main(String[] args) {
        System.out.println("╔═══════════════════════════════════════════════╗");
        System.out.println("║   DEPENDENCY INJECTION — Java                 ║");
        System.out.println("║   The Praetorian Quartermaster                ║");
        System.out.println("╚═══════════════════════════════════════════════╝\n");

        System.out.println("── PRODUCTION ───────────────────────────────────");
        new MilitaryCampaign(new ScrollMessenger(), new ConsoleLogger())
            .launchAttack("Gaul");

        System.out.println("\n── TESTING (mock) ───────────────────────────────");
        var mock = new MockMessenger();
        new MilitaryCampaign(mock, new ConsoleLogger()).launchAttack("Britannia");
        System.out.println("  Mock captured: " + mock.sent.size() + " messages ✓");

        System.out.println("\n\"Non ipse quaerit, accipit!\"");
    }
}

File: patterns/06_dependency_injection/python/officium.py
#!/usr/bin/env python3
"""
DEPENDENCY INJECTION — The Praetorian Quartermaster
"Non ipse quaerit, accipit" — He does not seek himself, he receives

Run: python3 officium.py
"""
from abc import ABC, abstractmethod
from typing import Protocol

class IMessenger(Protocol):
    def send(self, to: str, msg: str) -> None: ...

class ILogger(Protocol):
    def log(self, event: str) -> None: ...

class ScrollMessenger:
    def send(self, to: str, msg: str):
        print(f"  📜 Scroll to {to}: '{msg}'")

class SignalFireMessenger:
    def send(self, to: str, msg: str):
        print(f"  🔥 Signal fire to {to}: '{msg}'")

class MockMessenger:
    def __init__(self): self.sent = []
    def send(self, to: str, msg: str):
        self.sent.append((to, msg))
        print(f"  🧪 MOCK: to={to} msg='{msg}'")

class ConsoleLogger:
    def log(self, event: str):
        print(f"  [LOG] {event}")

class MilitaryCampaign:
    def __init__(self, messenger: IMessenger, logger: ILogger):
        self._messenger = messenger
        self._logger    = logger

    def launch_attack(self, target: str):
        self._logger.log(f"Campaign launched against {target}")
        self._messenger.send("Praetorian Guard", f"ADVANCE on {target}!")
        self._messenger.send("Supply Corps",     f"Move provisions to {target}")
        self._logger.log("Orders dispatched")

def main():
    print("╔═══════════════════════════════════════════════╗")
    print("║   DEPENDENCY INJECTION — Python               ║")
    print("║   The Praetorian Quartermaster                ║")
    print("╚═══════════════════════════════════════════════╝\n")

    print("── PRODUCTION (scroll messengers) ──────────────")
    MilitaryCampaign(ScrollMessenger(), ConsoleLogger()).launch_attack("Gaul")

    print("\n── SIGNAL FIRE VARIANT ─────────────────────────")
    MilitaryCampaign(SignalFireMessenger(), ConsoleLogger()).launch_attack("Britannia")

    print("\n── TESTING (mock — no real messages sent) ──────")
    mock = MockMessenger()
    MilitaryCampaign(mock, ConsoleLogger()).launch_attack("Germania")
    print(f"  Mock captured {len(mock.sent)} messages ✓")

    print("\n── KEY INSIGHT ─────────────────────────────────")
    print("MilitaryCampaign never creates its dependencies.")
    print("Swap ScrollMessenger for MockMessenger with ZERO code change.")
    print('"Non ipse quaerit, accipit!"')

if __name__ == "__main__":
    main()
